collections/src/main/scala/strawman/collection/mutable/ArrayDeque.scala

```diff
@@ -0,0 +1,519 @@
+    }
+    new ArrayDeque(arr, start = 0, end = n)
+  }
+
+  @inline def ensureSize(hint: Int) = if (hint > length && isResizeNecessary(hint)) resize(hint + 1)
+
+  override def length = end_-(start)
+
+  override def knownSize = length
+
+  override def isEmpty = start == end
+
+  override def nonEmpty = start != end
+
+  override def clone() = new ArrayDeque(array.clone(), start = start, end = end)
+
+  override def view: ArrayDequeView[A] = new ArrayDequeView(this)
+
+  override def iterableFactory: SeqFactory[ArrayDeque] = ArrayDeque
+
+  protected[this] def fromSpecificIterable(coll: collection.Iterable[A]): ArrayDeque[A] =
+    fromIterable(coll)
+
+  protected[this] def newSpecificBuilder(): Builder[A, ArrayDeque[A]] =
+    ArrayDeque.newBuilder()
+
+  /**
+    * Note: This does not actually resize the internal representation.
+    * See clearAndShrink if you want to also resize internally
+    */
+  override def clear() = {
+    while(nonEmpty) {
+      removeHeadAssumingNonEmpty()
+    }
+  }
+
+  /**
+    * clears this buffer and shrinks to @param size
+    *
+    * @param size
+    * @return
+    */
+  def clearAndShrink(size: Int = ArrayDeque.DefaultInitialSize): this.type = {
+    reset(array = ArrayDeque.alloc(size), start = 0, end = 0)
+    this
+  }
+
+  override def slice(from: Int, until: Int) = {
+    val n = length
+    val left = Math.max(0, Math.min(n, from))
+    val right = Math.max(0, Math.min(n, until))
+    val len = right - left
+    if (len <= 0) {
+      ArrayDeque.empty[A]
+    } else if (len >= n) {
+      clone()
+    } else {
+      val array2 = copySliceToArray(srcStart = left, dest = ArrayDeque.alloc(len), destStart = 0, maxItems = len)
+      new ArrayDeque(array2, start = 0, end = len)
+    }
+  }
+
+  override def sliding(window: Int, step: Int) = {
+    require(window > 0 && step > 0, s"window=$window and step=$step, but both must be positive")
+    val lag = if (window > step) window - step else 0
+    Iterator.range(start = 0, end = length - lag, step = step).map(i => slice(i, i + window))
+  }
+
+  override def grouped(n: Int) = sliding(n, n)
+
+  override def copyToArray[B >: A](dest: Array[B], destStart: Int, len: Int) =
+    copySliceToArray(srcStart = 0, dest = dest, destStart = destStart, maxItems = len)
+
+  override def className = "ArrayDeque"
+
+  override def toArray[B >: A: ClassTag] =
+    copySliceToArray(srcStart = 0, dest = new Array[B](length), destStart = 0, maxItems = length)
+
+  /**
+    * This is a more general version of copyToArray - this also accepts a srcStart unlike copyToArray
+    * This copies maxItems elements from this collections srcStart to dest's destStart
+    * If we reach the end of either collections before we could copy maxItems, we simply stop copying
+    *
+    * @param dest
+    * @param srcStart
+    * @param destStart
+    * @param maxItems
+    */
+  def copySliceToArray(srcStart: Int, dest: Array[_], destStart: Int, maxItems: Int): dest.type = {
```

Comment:
Why don’t you indicate the type of the elements in the destination array?

~~~ scala
def copySliceToArray[B >: A](srcStart: Int, dest: Array[B], destStart: Int, maxItems: Int): Array[B]
~~~

?

Reply:
I get this then:
```
[error] /Users/pbhowmick/Projects/collection-strawman/collections/src/main/scala/strawman/collection/mutable/ArrayDeque.scala:456: type mismatch;
[error]  found   : Array[AnyRef]
[error]  required: Array[Any]
[error] Note: AnyRef <: Any, but class Array is invariant in type T.
[error] You may wish to investigate a wildcard type such as `_ <: Any`. (SLS 3.2.10)
[error]     val array2 = copySliceToArray(srcStart = 0, dest = ArrayDeque.alloc(len), destStart = 0, maxItems = n)
[error]                                                                        ^
```

Reply:
Yeah, I saw that too. BTW, have you seen [my branch](https://github.com/scala/collection-strawman/compare/master...julienrf:array-dequeue) where I fixed all the compilation issues?